### PR TITLE
Request validation

### PR DESCRIPTION
**Pull request by rasapala**

### CODE REVIEW

File: src/test/capi_predict_validation_test.cpp [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+    preparePredictRequest(request,
+        {{inputName,
+            std::tuple<ovms::shape_t, ovms::Precision>{{1, 224, 224, 3}, testedPrecision}}},
+        {},     // data,
+        true);  // put buffer in InputTensorContent
+    auto status = instance->mockValidate(&request);
+    EXPECT_EQ(status, ovms::StatusCode::OK) << status.string();
+}
+
+class CAPIPredictValidationInputTensorContentNegative : public ::testing::Test {
```
Comment (atobiszei) [inline src/test/capi_predict_validation_test.cpp:447]: This class is not used anywhere

File: src/test/capi_predict_validation_test.cpp [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+            std::make_shared<ovms::TensorInfo>(inputName, testedPrecision, ovms::shape_t{1, 224, 224, 3}, ovms::Layout{"NHWC"})},
+    });
+    ON_CALL(*instance, getInputsInfo()).WillByDefault(ReturnRef(servableInputs));
+    ON_CALL(*instance, getBatchSize()).WillByDefault(Return(1));
+    ON_CALL(*instance, getModelConfig()).WillByDefault(ReturnRef(modelConfig));
+    preparePredictRequest(request,
+        {{inputName,
+            std::tuple<ovms::shape_t, ovms::Precision>{{1, 224, 224, 3}, testedPrecision}}},
+        {},     // data,
+        true);  // put buffer in InputTensorContent
```
Comment (atobiszei) [inline src/test/capi_predict_validation_test.cpp:442]: this comment is not relevant -> this is not KFS. Is it possible to skip this parameter? This applies to other places in this file as well.

File: src/test/capi_predict_validation_test.cpp [hunk 3]
```diff
@@ -0,0 +1,725 @@
...
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    Test,
+    CAPIPredictValidationPrecision,
+    ::testing::ValuesIn(SUPPORTED_CAPI_INPUT_PRECISIONS),
+    [](const ::testing::TestParamInfo<CAPIPredictValidationPrecision::ParamType>& info) {
+        return toString(info.param);
+    });
+
```
Comment (atobiszei) [inline src/test/capi_predict_validation_test.cpp:711]: We miss tests for incorrect buffer type. For now we probably should handle CPU buffers anyway, but it is possible to pass any value as input buffer type. The same is true for unsupported CAPI precisions and any arbitrary value that could be passed in OVMS_InferenceRequestAddInput.

File: src/test/capi_predict_validation_test.cpp [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+                std::tuple<ovms::shape_t, ovms::Precision>{{1, 223, 224, 3}, ovms::Precision::FP32}},
+            {"Input_U8_100:200_any_CN",
+                std::tuple<ovms::shape_t, ovms::Precision>{{101, 16}, ovms::Precision::U8}},
+        });
+    auto status = instance->mockValidate(&request);
+    EXPECT_EQ(status, ovms::StatusCode::INVALID_SHAPE) << status.string();
+}
+
+TEST_F(CAPIPredictValidationDynamicModel, RequestDimensionInRangeWrongTensorContent) {
+    //findKFSInferInputTensorContentInRawInputs(request, "Input_U8_100:200_any_CN")->clear();
```
Comment (atobiszei) [inline src/test/capi_predict_validation_test.cpp]: leftover comment here & in L655

File: src/test/capi_predict_validation_test.cpp [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+                std::tuple<ovms::shape_t, ovms::Precision>{{16, 300, 320, 3}, ovms::Precision::FP32}},
+            {"Input_U8_100:200_any_CN",
+                std::tuple<ovms::shape_t, ovms::Precision>{{98, 1}, ovms::Precision::U8}},
+        });
+    auto status = instance->mockValidate(&request);
+    EXPECT_EQ(status, ovms::StatusCode::INVALID_SHAPE) << status.string();
+}
+
+TEST_F(CAPIPredictValidationDynamicModel, RequestBatchNotInRangeSecondPosition) {
+    //prepareKFSInferInputTensor(request, "Input_U8_100:200_any_CN", {{100, 98}, "UINT8"});
```
Comment (atobiszei) [inline src/test/capi_predict_validation_test.cpp]: here

File: src/test/capi_predict_validation_test.cpp [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+            {"Input_U8_100:200_any_CN",
+                std::tuple<ovms::shape_t, ovms::Precision>{{100, 98}, ovms::Precision::U8}},
+        });
+    servableInputs["Input_U8_100:200_any_CN"] = std::make_shared<ovms::TensorInfo>("Input_U8_100:200_any_CN", ovms::Precision::U8, ovms::Shape{{100, 200}, {1, 5}}, ovms::Layout{"CN"});
+    auto status = instance->mockValidate(&request);
+    EXPECT_EQ(status, ovms::StatusCode::INVALID_BATCH_SIZE) << status.string();
+}
+
+TEST_F(CAPIPredictValidationDynamicModel, RequestDimensionNotInRangeSecondPosition) {
+    //prepareKFSInferInputTensor(request, "Input_FP32_any_224:512_224:512_3_NHWC", {{1, 223, 224, 3}, "FP32"});
```
Comment (atobiszei) [inline src/test/capi_predict_validation_test.cpp]: here

File: src/capi_frontend/capi_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+#include <unordered_map>
+
+#include "../logging.hpp"
+#include "../pocapi.hpp"
+#include "../precision.hpp"
+#include "../profiler.hpp"
+#include "../status.hpp"
+
+namespace ovms {
+Precision PrecisionToOvmsPrecision(const OVMSDataType& datatype) {
```
Comment (atobiszei) [inline src/capi_frontend/capi_utils.cpp:31]: Check pocapiinternal.[hc]pp

File: src/capi_frontend/capi_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    // {Precision::UNDEFINED, "UNDEFINED"}};
+
+    auto it = precisionMap.find(precision);
+    if (it == precisionMap.end()) {
+        it = precisionMap.find(Precision::INVALID);
+    }
+    return it->second;
+}
+
+std::string tensorShapeToString(const Shape& shape) {
```
Comment (dkalinowski) [inline src/capi_frontend/capi_utils.cpp:110]: This is not related to C-API but inside capi_utils.cpp

File: src/modelinstance.hpp [hunk 9]
```diff
@@ -41,6 +41,7 @@
 #include "tensorinfo.hpp"
 
 namespace ovms {
+class InferenceRequest;
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/modelinstance.hpp:44]: Forward declare only when we really need to (probably in modelinstance.cpp?)

File: src/precision.hpp [hunk 10]
```diff
@@ -44,7 +44,8 @@ enum class Precision {
     MIXED,
     Q78,
     BIN,  // TODO remove BIN,Q78, DYNAMIC, CUSTOM, MIXED
-    PRECISION_END
+    PRECISION_END,
+    INVALID
```
Comment (dkalinowski) [inline src/precision.hpp:48]: PRECISION_END should be the last code

Comment (dkalinowski) [reply src/precision.hpp:48]: Do we really want INVALID precision? The InferenceTensor API allows creating only "valid" tensors @atobiszei

File: src/predict_request_validation_utils.cpp [hunk 11]
```diff
@@ -342,6 +411,31 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
     }
     return StatusCode::OK;
 }
+template <>
+Status RequestValidator<ovms::InferenceRequest, ovms::InferenceTensor, const ovms::InferenceTensor*, shape_t>::checkBinaryBatchSizeMismatch(const InferenceTensor& proto, const Dimension& servableBatchSize, Status& finalStatus, Mode batchingMode, Mode shapeMode) const {
+    RequestShapeInfo<InferenceTensor, shape_t> rsi(proto);
+    if (proto.getBuffer()->getByteSize() <= 0) {
```
Comment (dkalinowski) [inline src/predict_request_validation_utils.cpp:417]: The error message state that the batch size is invalid - negative. You do not check batch size here but byte size, I guess thats typo?

Comment (dkalinowski) [reply src/predict_request_validation_utils.cpp:417]: If thats a typo then it looks like we do not have unit test for that, we probably should add it

File: src/test/test_utils.cpp [hunk 12]
```diff
@@ -374,6 +382,42 @@ void prepareKFSInferInputTensor(::KFSRequest& request, const std::string& name,
...
+    auto [shape, datatype] = inputInfo;
+    size_t elementsCount = 1;
+    for (auto const& dim : shape) {
+        ASSERT_GE(dim, 0);
+        elementsCount *= dim;
+    }
+
+    request.addInput(name.c_str(), datatype, shape.data(), shape.size());
+
+    if (!putBufferInInputTensorContent) {
```
Comment (dkalinowski) [inline src/test/test_utils.cpp:404]: In TFS/KFS this boolean was needed to decide where to put data - tensor content or val containers. Our C-API does not have multiple places for content storage therefore I think we should get rid of this boolean.

File: src/test/capi_predict_validation_test.cpp (hunk 1 above)
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp:446]: Removed.

File: src/test/capi_predict_validation_test.cpp (hunk 4 above)
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp]: Removed.

File: src/test/capi_predict_validation_test.cpp (hunk 5 above)
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp]: Removed.

File: src/test/capi_predict_validation_test.cpp (hunk 6 above)
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp]: Removed.

File: src/capi_frontend/capi_utils.cpp (hunk 7 above)
Comment (rasapala) [reply src/capi_frontend/capi_utils.cpp:32]: Done.

File: src/capi_frontend/capi_utils.cpp (hunk 8 above)
Comment (rasapala) [reply src/capi_frontend/capi_utils.cpp:111]: We can refactor this later. Currently implemented as in other APIs.

File: src/modelinstance.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (rasapala) [reply src/modelinstance.hpp]: Removed.

File: src/predict_request_validation_utils.cpp [hunk 13]
```diff
@@ -552,6 +664,22 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
     }
     return StatusCode::OK;
 }
+template <>
+Status RequestValidator<ovms::InferenceRequest, ovms::InferenceTensor, const ovms::InferenceTensor*, shape_t>::validateNumberOfShapeDimensions(const ovms::TensorInfo& inputInfo, const InferenceTensor& proto) const {
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/predict_request_validation_utils.cpp:668]: Fixed.

File: src/test/capi_predict_validation_test.cpp [hunk 14]
```diff
@@ -0,0 +1,719 @@
...
+#include "test_utils.hpp"
+
+using ::testing::NiceMock;
+using ::testing::Return;
+using ::testing::ReturnRef;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-variable"
+
+class MockModelInstance : public ovms::ModelInstance {
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp:36]: It would cause more problems than solve IMHO.

File: src/test/test_utils.cpp (hunk 12 above)
Comment (rasapala) [reply src/test/test_utils.cpp:404]: Removed.

File: src/predict_request_validation_utils.cpp (hunk 11 above)
Comment (rasapala) [reply src/predict_request_validation_utils.cpp:417]: Fixed.

File: src/precision.hpp (hunk 10 above)
Comment (rasapala) [reply src/precision.hpp:48]: Fixed.

File: src/capi_frontend/capi_utils.cpp [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            oss << shape[i].toString() << ",";
+        }
+        oss << shape[i].toString();
+    }
+    oss << ")";
+
+    return oss.str();
+}
+
+Status prepareConsolidatedTensorImpl(InferenceResponse* response, char*& bufferOut, const std::string& name, size_t size) {
```
Comment (atobiszei) [inline src/capi_frontend/capi_utils.cpp:71]: Why we need this? This is helper related to DAG's

File: src/statefulmodelinstance.hpp [hunk 16]
```diff
@@ -20,12 +20,12 @@
 #include <string>
 
 #include "global_sequences_viewer.hpp"
+#include "metric_config.hpp"
+#include "metric_registry.hpp"
```
Comment (atobiszei) [inline src/statefulmodelinstance.hpp:24]: Why we need to add headers here?
I assume cppclean complains here. The reason is most probably constructor definition inside hpp file. Please check if that will resolve this issue.

Both MetricRegistry & MetricConfig  are only passed as reference/pointer here so there should be no need for those headers.

File: src/test/prediction_service_test.cpp [hunk 17]
```diff
@@ -1152,7 +1152,7 @@ ovms::Precision getPrecisionFromResponse(KFSResponse& response, const std::strin
     size_t bufferId;
     auto status = getOutput(response, name, it, bufferId);
     EXPECT_TRUE(status.ok());
-    return ovms::KFSPrecisionToOvmsPrecision(it->datatype());
+    return ovms::KFSgetOVMSDataTypeAsPrecision(it->datatype());
```
Comment (atobiszei) [inline src/test/prediction_service_test.cpp:1155]: Why the change here and for tfs?

Comment (rasapala) [reply src/test/prediction_service_test.cpp:1155]: Fixed.

File: src/capi_frontend/capi_utils.cpp (hunk 15 above)
Comment (rasapala) [reply src/capi_frontend/capi_utils.cpp:71]: I can remove it.

File: src/predict_request_validation_utils.cpp (hunk 11 above)
Comment (rasapala) [reply src/predict_request_validation_utils.cpp]: Cannot set negative value for u int.

File: src/test/capi_predict_validation_test.cpp (hunk 2 above)
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp]: Removed.

File: src/test/capi_predict_validation_test.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp]: Fixed.

File: src/predict_request_validation_utils.cpp (hunk 11 above)
Comment (rasapala) [reply src/predict_request_validation_utils.cpp]: Tests can be added with binary input support.

File: src/test/capi_predict_validation_test.cpp [hunk 18]
```diff
@@ -0,0 +1,725 @@
...
+    EXPECT_EQ(status, ovms::StatusCode::INVALID_NO_OF_INPUTS) << status.string();
+}
+
+TEST_F(CAPIPredictValidation, RequestWrongInputName) {
+    request.removeInput("Input_U16_1_2_8_4_NCHW");
+    preparePredictRequest(request,
+        {{"BADInput_FP32_1_224_224_3_NHWC",
+            std::tuple<ovms::shape_t, ovms::Precision>{{1, 224, 224, 3}, ovms::Precision::FP32}}});
+    auto status = instance->mockValidate(&request);
+    EXPECT_EQ(status, ovms::StatusCode::INVALID_NO_OF_INPUTS) << status.string();
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp:101]: Added test for that.

File: src/predict_request_validation_utils.cpp [hunk 19]
```diff
@@ -196,20 +225,33 @@ Status RequestValidator<TFSRequestType, TFSInputTensorType, TFSInputTensorIterat
...
+    const std::string details = ss.str();
+    SPDLOG_DEBUG("[servable name: {} version: {}] Invalid number of inputs - {}", servableName, servableVersion, details);
+    return Status(StatusCode::INVALID_NO_OF_INPUTS, details);
+}
 
 template <>
 Status RequestValidator<TFSRequestType, TFSInputTensorType, TFSInputTensorIteratorType, TFSShapeType>::validateAndGetInput(const TFSRequestType& request, const std::string& name, TFSInputTensorIteratorType& it, size_t& bufferId) {
     it = request.inputs().find(name);
     if (it != request.inputs().end()) {
+        currentlyValidatedName = const_cast<std::string*>(&name);
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/predict_request_validation_utils.cpp:246]: Fixed.

File: src/test/capi_predict_validation_test.cpp (hunk 3 above)
Comment (rasapala) [reply src/test/capi_predict_validation_test.cpp:742]: Done.

File: src/buffer.cpp [hunk 20]
```diff
@@ -38,5 +38,11 @@ const void* Buffer::data() const {
 size_t Buffer::getByteSize() const {
     return byteSize;
 }
+const BufferType Buffer::getBufferType() const {
+    return bufferType;
+}
+const std::optional<uint32_t> Buffer::getDeviceId() const {
```
Comment (atobiszei) [inline src/buffer.cpp:44]: return reference instead of copy?

Comment (atobiszei) [reply src/buffer.cpp:44]: getBufferType can remain with copy since it's basic type but here we should have ref

File: src/capi_frontend/capi_utils.cpp (hunk 8 above)
Comment (atobiszei) [reply src/capi_frontend/capi_utils.cpp:56]: Use Shape::toString()

File: src/config.hpp [hunk 21]
```diff
@@ -240,6 +240,7 @@ class Config {
      *
      * @return uint
      */
+    uint32_t __maxSequenceNumber() const;  // TODO: Move to CLI parser when ready
```
Comment (atobiszei) [inline src/config.hpp:243]: should be removed?

File: src/capi_frontend/capi_utils.cpp [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+#include "../logging.hpp"
+#include "../pocapi.hpp"
+#include "../precision.hpp"
+#include "../profiler.hpp"
+#include "../shape.hpp"
+#include "../status.hpp"
+
+namespace ovms {
+
+size_t DataTypeSize(const OVMS_DataType& datatype) {
```
Comment (atobiszei) [inline src/capi_frontend/capi_utils.cpp:33]: This is already implemented in inferenceparameter.cpp. It should be moved out of there

File: src/serialization.hpp [hunk 23]
```diff
@@ -28,6 +28,7 @@
 #pragma GCC diagnostic pop
 
 #include "inferenceresponse.hpp"
+#include "inferencetensor.hpp"
```
Comment (atobiszei) [inline src/serialization.hpp:31]: REmove fwd declare from line 39

File: src/test/test_utils.hpp [hunk 24]
```diff
@@ -298,6 +308,24 @@ class ConstructorEnabledModelManager : public ovms::ModelManager {
     }
 };
 
+class RequestMockModelInstance : public ovms::ModelInstance {
```
Comment (atobiszei) [inline src/test/test_utils.hpp:311]: Rename to MockedMetadataModelInstance? Or sth similar to know from name of the class what is mocked

File: src/predict_request_validation_utils.cpp [hunk 25]
```diff
@@ -612,6 +770,15 @@ bool RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorTyp
     }
     return false;
 }
+template <>
+bool RequestValidator<ovms::InferenceRequest, InferenceTensor, const InferenceTensor*, shape_t>::checkIfBinaryInputUsed(const InferenceTensor& proto, const std::string inputName) const {
+    // TODO no strig no bytes ?
```
Comment (atobiszei) [inline src/predict_request_validation_utils.cpp:775]: leave TODO, remove commented code

File: src/predict_request_validation_utils.cpp [hunk 26]
```diff
@@ -580,6 +726,19 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
     }
     return StatusCode::OK;
 }
+template <>
```
Comment (atobiszei) [inline src/predict_request_validation_utils.cpp:729]: C-API does not use protobuf so it would be good to not use proto as a parameter name to not confuse

File: src/predict_request_validation_utils.cpp [hunk 27]
```diff
@@ -518,6 +631,24 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
...
+Status RequestValidator<ovms::InferenceRequest, InferenceTensor, const InferenceTensor*, shape_t>::validateTensorContentSize(const InferenceTensor& proto, ovms::Precision expectedPrecision, size_t bufferId) const {
```
Comment (atobiszei) [inline src/predict_request_validation_utils.cpp:635]: proto->tensor

File: src/predict_request_validation_utils.cpp [hunk 28]
```diff
@@ -552,6 +683,22 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
     }
     return StatusCode::OK;
 }
+template <>
+Status RequestValidator<ovms::InferenceRequest, InferenceTensor, const InferenceTensor*, shape_t>::validateNumberOfShapeDimensions(const ovms::TensorInfo& inputInfo, const InferenceTensor& proto) const {
```
Comment (atobiszei) [inline src/predict_request_validation_utils.cpp:687]: proto->tensor

File: src/predict_request_validation_utils.cpp [hunk 29]
```diff
@@ -518,6 +631,24 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
     }
     return StatusCode::OK;
 }
+template <>
+Status RequestValidator<ovms::InferenceRequest, InferenceTensor, const InferenceTensor*, shape_t>::validateTensorContentSize(const InferenceTensor& proto, ovms::Precision expectedPrecision, size_t bufferId) const {
+    size_t expectedValueCount = 1;
+    for (size_t i = 0; i < proto.getShape().size(); i++) {
+        expectedValueCount *= proto.getShape()[i];
+    }
+    if (proto.getBuffer()->getByteSize()) {
```
Comment (atobiszei) [inline src/predict_request_validation_utils.cpp:640]: what happens when tensor claims that its buffer bytesize is 0?

Comment (atobiszei) [reply src/predict_request_validation_utils.cpp:640]: I assume that at this point we already validated that the InferenceTensor has Buffer attached to it?

File: src/predict_request_validation_utils.cpp [hunk 30]
```diff
@@ -222,8 +265,42 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
...
+        const BufferType bufType = buffer->getBufferType();
+        if (bufType < BufferType::OVMS_BUFFERTYPE_CPU || bufType > BufferType::OVMS_BUFFERTYPE_HDDL) {
+            std::stringstream ss;
+            ss << "Required input: " << name;
+            const std::string details = ss.str();
+            SPDLOG_DEBUG("[servable name: {} version: {}] Has invalid buffer type for input with specific name - {}", servableName, servableVersion, details);
+            return Status(StatusCode::INVALID_BUFFER_TYPE, details);
+        }
+
+        if (buffer->getBufferType() == BufferType::OVMS_BUFFERTYPE_CPU && buffer->getDeviceId() != 0) {
```
[the remark this answers is not in the document]
Comment (atobiszei) [reply src/predict_request_validation_utils.cpp:292]: Be sure to add test for that

File: src/test/test_utils.cpp [hunk 31]
```diff
@@ -374,6 +383,40 @@ void prepareKFSInferInputTensor(::KFSRequest& request, const std::string& name,
...
+
+    std::string* content = new std::string("");
+    if (data.size() == 0) {
+        content->assign(elementsCount * ovms::DataTypeSize(datatype), '1');
+    } else {
+        content->resize(elementsCount * ovms::DataTypeSize(datatype));
+        std::memcpy(content->data(), data.data(), content->size());
+    }
+
+    size_t dataSize = data.size();
```
[the remark this answers is not in the document]
Comment (atobiszei) [reply src/test/test_utils.cpp:415]: This is now even more relevant since you delete the ptr straight away

File: src/test/test_utils.hpp [hunk 32]
```diff
@@ -161,14 +161,14 @@ void prepareKFSInferInputTensor(::KFSRequest& request, const std::string& name,
     const std::vector<float>& data = {}, bool putBufferInInputTensorContent = false);
 
 void prepareCAPIInferInputTensor(ovms::InferenceRequest& request, const std::string& name, const std::tuple<ovms::shape_t, OVMS_DataType>& inputInfo,
-    const std::vector<float>& data = {}, bool badSize = false, BufferType bufferType = BufferType::OVMS_BUFFERTYPE_CPU, uint32_t deviceId = 0);
+    const std::vector<float>& data = {}, bool badSize = false, BufferType bufferType = BufferType::OVMS_BUFFERTYPE_CPU, std::optional<uint32_t> deviceId = std::nullopt);
```
Comment (atobiszei) [inline src/test/test_utils.hpp:164]: use reference to optional

File: src/capi_frontend/capi_utils.hpp [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+// limitations under the License.
+//*****************************************************************************
+#pragma once
+#include <string>
+
+#include "../pocapi.hpp"
+#include "../precision.hpp"
+
+namespace ovms {
+class Status;
```
Comment (atobiszei) [inline src/capi_frontend/capi_utils.hpp:23]: Cpp does not complain about unnecessary fwd declares of Status & Inference response? 2 includes (pocapi.hpp & precision.hpp) are most likely not needed now?

File: src/predict_request_validation_utils.cpp [hunk 34]
```diff
@@ -222,8 +265,42 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
...
+    SPDLOG_DEBUG("[servable name: {} version: {}] Missing input with specific name - {}", servableName, servableVersion, details);
+    return Status(StatusCode::INVALID_MISSING_INPUT, details);
+}
+template <>
+Status RequestValidator<ovms::InferenceRequest, InferenceTensor, const InferenceTensor*, shape_t>::validateAndGetInput(const InferenceRequest& request, const std::string& name, const InferenceTensor*& it, size_t& bufferId) {
+    if (request.getInput(name.c_str(), &it) != StatusCode::NONEXISTENT_TENSOR) {
+        currentlyValidatedName = &name;
+        const Buffer* buffer = it->getBuffer();
+        const BufferType bufType = buffer->getBufferType();
+        if (bufType < BufferType::OVMS_BUFFERTYPE_CPU || bufType > BufferType::OVMS_BUFFERTYPE_HDDL) {
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/predict_request_validation_utils.cpp:284]: Why was it resolved?

File: src/predict_request_validation_utils.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/predict_request_validation_utils.cpp]: Why was this resolved?

File: src/predict_request_validation_utils.cpp [hunk 35]
```diff
@@ -222,8 +265,45 @@ Status RequestValidator<KFSRequest, KFSTensorInputProto, KFSInputTensorIteratorT
...
+            // Remove this when other buffer types are supported
+            if (bufType != BufferType::OVMS_BUFFERTYPE_CPU) {
```
Comment (dkalinowski) [inline src/predict_request_validation_utils.cpp:286]: This will never execute when buffer type is hddl. It will not enter if branch in 284. It probably should not be nested if but one after another

Comment (dkalinowski) [reply src/predict_request_validation_utils.cpp:286]: Also add test that specifying hddl (and other unsupported types) fails validation. It must be mising since this code passed

**Review by dkalinowski — APPROVED** [review]

File: src/test/test_utils.cpp [hunk 36]
```diff
@@ -374,6 +384,40 @@ void prepareKFSInferInputTensor(::KFSRequest& request, const std::string& name,
...
+        elementsCount *= dim;
+    }
+
+    request.addInput(name.c_str(), datatype, shape.data(), shape.size());
+
+    size_t dataSize = elementsCount * ovms::DataTypeToByteSize(datatype);
+    if (decrementBufferSize)
+        dataSize -= decrementBufferSize;
+
+    std::string* content = new std::string("");
```
Comment (atobiszei) [inline src/test/test_utils.cpp:410]: Memory leak -> can be fixed later but please leave TODO

**Review by atobiszei — APPROVED** [review]
